src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from typing import ClassVar, Generator, Pattern
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.option.global_options import GlobalOptions
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+_pex_target_regexp: str = r"""
+(?# optional path, one level with dot-separated words)
+(?:(?P<path>(?:\w[.0-9_-]?)+) /)?
+
+(?# binary name, with .pex file extension)
+(?P<name>(?:\w[.0-9_-]?)+) \.pex$
+"""
+
+
+@dataclass(frozen=True)
+class ParsedDockerfile:
+    commands: tuple[Command, ...]
+    pex_target_regexp: ClassVar[Pattern]
+
+    def __post_init__(self):
+        # Compile regexp when creating the first instance, and store it on the class.
```

Comment:
It's not obvious to me that this is thread-safe. Assignment to a simple variable is atomic in Python, but there is no clear answer on whether assignment to a member (which involves the object's member dictionary) is atomic. In practice it probably depends on the implementation. 

My apologies for making the regex issue more complicated than it needed to be, but I'm now thinking we should do the simple thing you did all along and compile at module load time, storing the compiled regex directly in the module. In v2 there are so few regexes that this is not going to be a performance issue. Again, sorry for spinning you in circles on this...

Reply:
Well, now we've explored the alternatives and can motivate the (small) cost of load-time compiling the regexp. I can leave a short comment about it so it is not lost and forgotten. 
Running in circles is common for me 🐶